---
title: "R Notebook"
output: html_notebook
---


```{r}

library(Seurat)
library(SeuratObject)
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/tools/seurat_functions.R")
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/tools/seurat_integration_functions.R")

data.path <- "E:/STOmics/seurat/raw/"
obj.data <- Read10X(data.path, gene.column = 1)
obj <- CreateSeuratObject(counts = obj.data, project = "Opossum_V1_Stereo-seq")

mapping_path <- "E:/_genomes/Opossum_Mouse_Genes.txt"
obj <- MapGenes(obj, mapping_path, use_ids = TRUE)

obj$species <- "Opossum"
X <- read.csv(paste(data.path, "x_coords.csv", sep = ""), header = FALSE)
X <- (X - max(X)) * -1
obj <- AddMetaData(obj, X$V1, "X")
Y <- read.csv(paste(data.path, "y_coords.csv", sep = ""), header = FALSE)
Y <- (Y - max(Y)) * -1
obj <- AddMetaData(obj, Y$V1, "Y")

cents.df <- data.frame(X, Y)
colnames(cents.df) <- c("X", "Y")
rownames(cents.df) <- colnames(obj)
cents <- CreateCentroids(cents.df)
fov <- CreateFOV(
  cents, 
  type = "centroids",
  assay = "RNA",
  key = Key("FOV", quiet = TRUE)
)

fov <- fov[Cells(obj)]

obj[["FOV"]] <- fov

```


```{r, fig.width=10}

ImageDimPlot(obj, cols = "red")
ImageFeaturePlot(obj, "nCount_RNA", max.cutoff = 1000)
# ImageFeaturePlot(obj, "", max.cutoff = 10, axes = TRUE)

```


```{r}

saveRDS(obj, "E:/STOmics/seurat/opossum_stereoseq_all.rds")

```

