---
title: "R Notebook"
output: html_notebook
---


```{r}

library(Seurat)
library(SeuratObject)
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/tools/seurat_functions.R")
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/tools/seurat_integration_functions.R")

data.path <- "E:/STOmics/seurat/Opossum/raw/"
obj.data <- Read10X(data.path, gene.column = 1)
obj <- CreateSeuratObject(counts = obj.data, project = "Opossum_V1_Stereo-seq")

mapping_path <- "E:/_genomes/Opossum_Mouse_Genes.txt"
obj <- MapGenes(obj, mapping_path, use_ids = TRUE)

obj$species <- "Opossum"
X <- read.csv(paste(data.path, "x_coords.csv", sep = ""), header = FALSE) * -1
obj <- AddMetaData(obj, X$V1, "X")
Y <- read.csv(paste(data.path, "y_coords.csv", sep = ""), header = FALSE) * -1
obj <- AddMetaData(obj, Y$V1, "Y")

crops <- list(theta = c(-5, -7, -38), 
              x1 = c(-9500, -7000, -125), 
              x2 = c(-7500, -5000, 1045), 
              y1 = c(-14300, -14600, -5650), 
              y2 = c(-12800, -13050, -7000))

# Define the rotation angle in degrees
rotation_angle <- -5  # Rotate by 30 degrees
theta <- rotation_angle * pi / 180  # Convert to radians

# Get current coordinates
coords <- as.matrix(obj@meta.data[, c("X", "Y")])

# Define rotation matrix
rotation_matrix <- matrix(
  c(cos(theta), -sin(theta),
    sin(theta),  cos(theta)),
  nrow = 2
)

# Apply rotation
rotated_coords <- coords %*% rotation_matrix

# Update metadata
obj$X_rotated <- rotated_coords[, 1]
obj$Y_rotated <- rotated_coords[, 2]

cents.df <- data.frame(X, Y)
colnames(cents.df) <- c("X", "Y")
rownames(cents.df) <- colnames(obj)

cents.df$X <- obj$X_rotated
cents.df$Y <- obj$Y_rotated

cents <- CreateCentroids(cents.df)
fov <- CreateFOV(
  cents, 
  type = "centroids",
  assay = "RNA",
  key = Key("FOV", quiet = TRUE)
)

fov <- fov[Cells(obj)]

obj[["FOV"]] <- fov

x1 <- -9500
x2 <- -7500
y1 <- -14300
y2 <- -12800
extension <- (y2 - y1) * 0.1
column.coords <- Crop(obj[["FOV"]], x = c(x1, x2), y = c(y1 - extension, y2), coords = "plot")
obj[["COL1"]] <- column.coords

```


```{r, fig.width=10}

ImageDimPlot(obj, fov = "COL1", cols = "red", size = 3)
# ImageFeaturePlot(obj, "nCount_RNA", max.cutoff = 1000)
# ImageFeaturePlot(obj, "", max.cutoff = 10, axes = TRUE)

```






```{r}

obj.ctx <- subset(obj, cells = Cells(obj[["COL1"]]))
saveRDS(obj.ctx, "E:/STOmics/seurat/Opossum/opossum_stereoseq_ctx_COL1_EXTEXT.rds")

```

