---
title: "R Notebook"
output: html_notebook
---


```{r}

library(Seurat)
library(SeuratDisk)
library(reticulate)
library(scrubletR)
library(ggplot2)
library(patchwork)
library(dplyr)
library(data.table)
library(clustree)
library(reshape2)
library(tidyr)
library(gridExtra)
library(stringr)
library(plyr)
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/tools/seurat_functions.R")
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/tools/seurat_integration_functions.R")
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/xgboost/xgboost_train.R")
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/xgboost/plottingFxns.R")
colors_list <- GetColors()

data.path <- "E:/STOmics/seurat/Opossum/raw/"
obj.data <- Read10X(data.path, gene.column = 1)
obj.section <- CreateSeuratObject(counts = obj.data, project = "Opossum_V1_Stereo-seq")

mapping_path <- "E:/_genomes/Opossum_Mouse_Genes.txt"
obj.section <- MapGenes(obj.section, mapping_path, use_ids = TRUE)

obj.section$species <- "Opossum"
X <- read.csv(paste(data.path, "x_coords.csv", sep = ""), header = FALSE)
X <- (X - max(X)) * -1
obj.section <- AddMetaData(obj.section, X$V1, "X")
Y <- read.csv(paste(data.path, "y_coords.csv", sep = ""), header = FALSE)
Y <- (Y - max(Y)) * -1
obj.section <- AddMetaData(obj.section, Y$V1, "Y")

cents.df <- data.frame(X, Y)
colnames(cents.df) <- c("X", "Y")
rownames(cents.df) <- colnames(obj.section)
cents <- CreateCentroids(cents.df)
fov <- CreateFOV(
  cents, 
  type = "centroids",
  assay = "RNA",
  key = Key("FOV", quiet = TRUE)
)

section.cells <- read.csv("E:/STOmics/seurat/Opossum/selected_barcodes_section1.csv", header = FALSE)

fov <- fov[as.character(section.cells$V1)]

obj.section[["FOV"]] <- fov

```


```{r, fig.width=10}

ImageDimPlot(obj.section, cols = "red")
ImageFeaturePlot(obj.section, "nCount_RNA", max.cutoff = 1000)
# ImageFeaturePlot(obj, "", max.cutoff = 10, axes = TRUE)

```


```{r}

obj.section <- ClusterSCT(obj.section, 0.1)

```


```{r, fig.width=10}

ImageDimPlot(obj.section,  group.by = "SCT_snn_res.0.1", size = 2)

```


```{r}

obj.glutamatergic <- readRDS("E:/Transcriptomics_V1/Opossum/seurat/opossum_v1_glutamatergic_processed.rds")
obj.glutamatergic$class <- "glutamatergic"
obj.gabaergic <- readRDS("E:/Transcriptomics_V1/Opossum/seurat/opossum_v1_gabaergic_processed.rds")
obj.gabaergic$class <- "gabaergic"
obj.nonneuronal <- readRDS("E:/Transcriptomics_V1/Opossum/seurat/opossum_v1_nonneuronal_processed.rds")

obj <- merge(obj.glutamatergic, y = c(obj.gabaergic, obj.nonneuronal))

rm(obj.glutamatergic)
rm(obj.gabaergic)
rm(obj.nonneuronal)

obj <- obj[, sample(colnames(obj), ncol(obj)/2, replace = FALSE)]

```


```{r}

cell_mask <- Reduce(intersect, list(WhichCells(obj.section, expression = nFeature_RNA > 50),
                                    WhichCells(obj.section, expression = nCount_RNA > 100)))
  
gene_mask <- rownames(obj.section)[Matrix::rowSums(obj.section[["RNA"]]@counts > 0) > 8]

obj.section <- subset(obj.section, features = gene_mask, cells = cell_mask)

```


```{r}

obj$method <- "snRNA-seq"
obj.section$method <- "Stereo-seq"

```


```{r}

df <- setNames(data.frame(matrix(ncol = 2, nrow = 0)), c("cell", "subclass_nn"))
cell.names.sp <- colnames(obj.section)

for (i in 5:ceiling(ncol(obj.section) / 1000)) {
  
  # Sample names unless there's too few remaining
  if (length(cell.names.sp) >= ncol(obj)) {
    curr.names <- sample(cell.names.sp, 1000)
    cell.names.sp <- setdiff(cell.names.sp, curr.names)
  } else { curr.names <- cell.names.sp }
  
  # Subset object and integrate
  obj.section.i <- obj.section[, curr.names]
  obj.i <- IntegrateObjects(obj, obj.section.i, resolutions = c(0.5), subsample = FALSE)
  
  # Make NN subclass assignments
  obj.i <- LabelByNearestNeighbors(obj.i, "subclass", fraction = 0.25, n.neighbors = 100)
  
  # Grab the metadata
  curr.df <- data.frame(cell = colnames(obj.i)[obj.i$method == "Stereo-seq"], 
                        subclass_nn = as.character(obj.i$subclass_nn[obj.i$method == "Stereo-seq"]))
  print(table(curr.df$subclass_nn)["None"] / sum(table(curr.df$subclass_nn)))
  df <- rbind(df, curr.df)
  
}

saveRDS(df, "E:/STOmics/seurat/Opossum/opossum_section_subclass_labels.rds")

```


```{r}

DimPlot(obj.i, group.by = "subclass_nn")

```



```{r}

# Read the data
df <- readRDS("E:/STOmics/seurat/Opossum/opossum_section_subclass_labels.rds")

# Assign to spatial Seurat object
obj.section$subclass_nn <- NA
obj.section$subclass_nn[df$cell] <- df$subclass_nn

```


```{r, fig.width=15}

obj.section.plot <- subset(obj.section, subclass_nn %in% c("IT_A", "IT_B", "IT_C", "L5NP", "L5PT", "IT_D", "L6CT", "L6b", "OD"))

Idents(obj.section.plot) <- "subclass_nn"
subclasses.opossum <- c("IT_A", "IT_B", "IT_C", "L5NP", "L5PT", "IT_D", "L6CT", "L6b", 
                         "OD")
colors <- hcl.colors(length(subclasses.opossum), palette = "dynamic")
colors[2] <- "#FF6347"
levels(obj.section.plot) <- subclasses.opossum
p <- ImageDimPlot(obj.section.plot, cols = colors, size = 2.5)
ggsave("E:/Opossum_Paper/Figure 3/Opossum_Section.svg", plot = p)

```










