---
title: "R Notebook"
output: html_notebook
---

```{r}

library(Seurat)
library(SeuratDisk)
library(reticulate)
library(scrubletR)
library(ggplot2)
library(patchwork)
library(dplyr)
library(data.table)
library(clustree)
library(reshape2)
library(tidyr)
library(gridExtra)
library(stringr)
library(plyr)
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/tools/seurat_functions.R")
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/tools/seurat_integration_functions.R")
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/xgboost/xgboost_train.R")
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/xgboost/plottingFxns.R")
colors_list <- GetColors()

obj.glutamatergic <- readRDS("E:/Transcriptomics_V1/Opossum/seurat/opossum_v1_glutamatergic_processed.rds")
obj.gabaergic <- readRDS("E:/Transcriptomics_V1/Opossum/seurat/opossum_v1_gabaergic_processed.rds")
obj.nonneuronal <- readRDS("E:/Transcriptomics_V1/Opossum/seurat/opossum_v1_nonneuronal_processed.rds")

obj <- merge(obj.glutamatergic, y = c(obj.gabaergic, obj.nonneuronal))

obj.ctx <- readRDS("E:/STOmics/seurat/opossum_stereoseq_ctx_1.rds")

```


```{r}

VlnPlot(obj.ctx, "nFeature_RNA")
VlnPlot(obj.ctx, "nCount_RNA")
summary(obj.ctx$nFeature_RNA)
summary(obj.ctx$nCount_RNA)

```


```{r, fig.width=12}

ImageDimPlot(obj.ctx, size = 2.5)

```


```{r}

cell_mask <- Reduce(intersect,list(WhichCells(obj.ctx, expression = nFeature_RNA > 50),
                                   WhichCells(obj.ctx, expression = nCount_RNA > 100)))
  
gene_mask <- rownames(obj.ctx)[Matrix::rowSums(obj.ctx[["RNA"]]@counts > 0) > 8]

obj.ctx <- subset(obj.ctx, features = gene_mask, cells = cell_mask)

```


```{r}

VlnPlot(obj.ctx, "nFeature_RNA")
VlnPlot(obj.ctx, "nCount_RNA")
summary(obj.ctx$nFeature_RNA)
summary(obj.ctx$nCount_RNA)

```


```{r, fig.width=12}

ImageDimPlot(obj.ctx, size = 2.5)

```


```{r}

obj <- ClusterSCT(obj, resolutions = 1)
obj.ctx <- ClusterSCT(obj.ctx, resolutions = c(0.5, 1))

```


```{r}

opossum.subclass.levels <- c("IT_A", "IT_C", "IT_B", "IT_D", "L5NP", "L5PT", "L6CT", "L6b", "Pvalb", "Sst", "Vip", "Lamp5", "Frem1", "OD", "OPC", "Astro", "Micro", "Endo")

DimPlot(obj, group.by = "subclass", cols = colors_list[opossum.subclass.levels]) + NoLegend() + xlim(-20, 20) + ylim(-20, 20) + coord_equal()
DimPlot(obj.ctx, group.by = "SCT_snn_res.0.5") + NoLegend() + xlim(-8, 8) + ylim(-8, 8) + coord_equal()
DimPlot(obj.ctx, group.by = "SCT_snn_res.1") + NoLegend() + xlim(-8, 8) + ylim(-8, 8) + coord_equal()

```


```{r, fig.width=12}

ImageDimPlot(obj.ctx, group.by = "SCT_snn_res.0.5", size = 2.5)
ImageDimPlot(obj.ctx, group.by = "SCT_snn_res.1", size = 2.5)

```


```{r}

obj.ctx <- MapObject(obj, obj.ctx, "class", do.norm = FALSE)

```


```{r, fig.width=12}

ImageDimPlot(obj.ctx, group.by = "predicted.class", size = 2.5)

```


```{r}

saveRDS(obj.ctx, "E:/STOmics/seurat/opossum_stereoseq_ctx_1_clustered_mapped.rds")
# obj.ctx <- readRDS("E:/STOmics/seurat/opossum_stereoseq_ctx_1_clustered.rds")

```


```{r}

obj.ctx.glutamatergic <- subset(obj.ctx, predicted.class == "glutamatergic")
obj.ctx.glutamatergic <- ClusterSCT(obj.ctx.glutamatergic, resolutions = 1)
obj.glutamatergic <- ClusterSCT(obj.glutamatergic, resolutions = 1)
obj.ctx.glutamatergic <- MapObject(obj.glutamatergic, obj.ctx.glutamatergic, "subclass", do.norm = FALSE)

```


```{r}

saveRDS(obj.glutamatergic, "E:/STOmics/seurat/opossum_glutamatergic_reference.rds")
saveRDS(obj.ctx.glutamatergic, "E:/STOmics/seurat/opossum_stereoseq_glutamatergic_1_clustered_mapped.rds")

obj.glutamatergic <- readRDS("E:/STOmics/seurat/opossum_glutamatergic_reference.rds")
obj.ctx.glutamatergic <- readRDS("E:/STOmics/seurat/opossum_stereoseq_glutamatergic_1_clustered_mapped.rds")

```


```{r, fig.width=12}

ImageFeaturePlot(obj.ctx.glutamatergic, "Cux2", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic, "Rorb", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)

```


```{r, fig.width=12}

Idents(obj.ctx.glutamatergic) <- "predicted.subclass"
levels(obj.ctx.glutamatergic) <- c("IT_A", "IT_B", "IT_C", "L5PT", "L6CT", "L5NP")
ImageDimPlot(obj.ctx.glutamatergic, size = 2.5, cols = c("#A0A0A0", "#FF7F50", "#FFA07A", "#128e27", "#1E90FF", "#C934EB"))
ImageDimPlot(obj.ctx.glutamatergic, size = 2.5, group.by = "SCT_snn_res.1")
PlotClusterImageDim(obj.ctx.glutamatergic)

```


```{r}

DimPlot(obj.ctx.glutamatergic, group.by = "SCT_snn_res.1", label = TRUE) + NoLegend() + xlim(-8, 8) + ylim(-8, 8) + coord_equal()
Idents(obj.ctx.glutamatergic) <- "predicted.subclass"
levels(obj.ctx.glutamatergic) <- c("IT_A", "IT_B", "IT_C", "L5PT", "L6CT", "L5NP")
DimPlot(obj.ctx.glutamatergic, cols = c("#A0A0A0", "#FF7F50", "#FFA07A", "#128e27", "#1E90FF", "#C934EB")) + xlim(-8, 8) + ylim(-8, 8) + coord_equal()

```


```{r, fig.width=6}

column.coords <- Crop(obj.ctx.glutamatergic[["CTX"]], x = c(6750, 8500), y = c(10000, 12000), coords = "tissue")
obj.ctx.glutamatergic[["COL"]] <- column.coords
levels(obj.ctx.glutamatergic) <- c("IT_A", "IT_B", "IT_C", "L5PT", "L6CT", "L5NP")
ImageDimPlot(obj.ctx.glutamatergic, fov = "COL", size = 3, cols = c("#A0A0A0", "#FF7F50", "#FFA07A", "#128e27", "#1E90FF", "#C934EB"))

```


```{r}

obj.ctx.glutamatergic.COL <- subset(obj.ctx.glutamatergic, cells = Cells(obj.ctx.glutamatergic[["COL"]]))
obj.ctx.glutamatergic.COL <- ClusterSCT(obj.ctx.glutamatergic.COL, resolutions = 1)

```


```{r}

obj.ctx.glutamatergic.COL <- MapObject(obj.glutamatergic, obj.ctx.glutamatergic.COL, "subclass", do.norm = FALSE)

```


```{r}

Idents(obj.ctx.glutamatergic.COL) <- "predicted.subclass"
levels(obj.ctx.glutamatergic.COL) <- c("IT_A", "IT_B", "IT_C", "L5PT", "L6CT", "L5NP")
ImageDimPlot(obj.ctx.glutamatergic.COL, fov = "COL", size = 3, cols = c("#A0A0A0", "#FF7F50", "#FFA07A", "#128e27", "#1E90FF", "#C934EB"))
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Cux2", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)

```


```{r}

DimPlot(obj.ctx.glutamatergic.COL, group.by = "SCT_snn_res.1", label = TRUE) + NoLegend() + xlim(-8, 8) + ylim(-8, 8) + coord_equal()
DimPlot(obj.ctx.glutamatergic.COL, group.by = "predicted.subclass", cols = c("#A0A0A0", "#FF7F50", "#FFA07A", "#128e27", "#1E90FF")) + xlim(-8, 8) + ylim(-8, 8) + coord_equal()
FeaturePlot(obj.ctx.glutamatergic.COL, "Foxp2") + xlim(-8, 8) + ylim(-8, 8) + coord_equal()

```


```{r}

Idents(obj.ctx.glutamatergic.COL) <- "SCT_snn_res.1"
ImageDimPlot(obj.ctx.glutamatergic.COL, fov = "COL", size = 3)
PlotClusterImageDim(obj.ctx.glutamatergic.COL)

```


```{r}

markers.6 <- FindMarkers(obj.ctx.glutamatergic.COL, ident.1 = 6, only.pos = TRUE)

```


```{r}

length(intersect(VariableFeatures(obj.ctx.glutamatergic.COL), VariableFeatures(obj.glutamatergic)))
length(intersect(rownames(obj.ctx.glutamatergic.COL), VariableFeatures(obj.glutamatergic)))

snRNAseq.vfs <- intersect(rownames(obj.ctx.glutamatergic.COL), VariableFeatures(obj.glutamatergic))

VariableFeatures(obj.ctx.glutamatergic.COL) <- snRNAseq.vfs
obj.ctx.glutamatergic.COL.VFs <- obj.ctx.glutamatergic.COL %>%
                                   RunPCA(npcs = 30, features = snRNAseq.vfs, verbose = FALSE) %>%
                                   FindNeighbors(reduction = "pca", dims = 1:30, verbose = FALSE) %>%
                                   FindClusters(algorithm = 4, method = "igraph") %>%
                                   RunUMAP(reduction = "pca", dims = 1:30, verbose = FALSE)

```


```{r}

DimPlot(obj.ctx.glutamatergic.COL.VFs, group.by = "SCT_snn_res.1", label = TRUE) + NoLegend() + xlim(-8, 8) + ylim(-8, 8) + coord_equal()
DimPlot(obj.ctx.glutamatergic.COL.VFs, group.by = "predicted.subclass") + xlim(-8, 8) + ylim(-8, 8) + coord_equal()

```


```{r}

library(Seurat)
library(dplyr)
library(ComplexHeatmap)

# Compute DE genes for obj1
de_snRNAseq <- FindAllMarkers(
  object = obj.glutamtergic,
  only.pos = TRUE,
  min.pct = pct.threshold,
  logfc.threshold = logfc.threshold,
  test.use = 'wilcox'
) %>% filter(p_val_adj < pval.threshold)

# Compute DE genes for obj2
Idents(obj.ctx.glutamatergic.COL) <- ""
de_Stereoseq <- FindAllMarkers(
  object = obj.ctx.glutamatergic.COL,
  only.pos = TRUE,
  min.pct = pct.threshold,
  logfc.threshold = logfc.threshold,
  test.use = 'wilcox'
) %>% filter(p_val_adj < pval.threshold)

# Updated function to take precomputed DE genes as input
ComputeOverlapHeatmap <- function(de_obj1, de_obj2, ident1 = 'subclass', ident2 = 'SCT_snn_res.1') {
  library(Seurat)
  library(dplyr)
  library(ComplexHeatmap)

  # Create lists of DE genes
  de_list1 <- split(de_obj1$gene, de_obj1[[ident1]])
  de_list2 <- split(de_obj2$gene, de_obj2[[ident2]])

  # Create overlap matrix
  overlap_matrix <- matrix(0, nrow = length(de_list1), ncol = length(de_list2),
                           dimnames = list(names(de_list1), names(de_list2)))
  for (i in seq_along(de_list1)) {
    for (j in seq_along(de_list2)) {
      overlap <- length(intersect(de_list1[[i]], de_list2[[j]]))
      total <- length(unique(c(de_list1[[i]], de_list2[[j]])))
      overlap_matrix[i, j] <- overlap / total
    }
  }

  # Plot heatmap
  Heatmap(overlap_matrix, name = 'Fraction Overlap', cluster_rows = TRUE, cluster_columns = TRUE)
}

```


```{r}

Idents(obj.glutamatergic) <- "subclass"
markers.IT_B <- FindMarkers(obj.glutamatergic, ident.1 = "IT_B", only.pos = TRUE)

```


```{r}

# Look for L4-specific genes
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Rorb", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Scnn1a", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Rspo1", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Thsd7a", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
# ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Ntsr2", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Cux2", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Kcnk2", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
# ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Fam84b", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Tshz2", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Sema3e", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)

```


```{r}

ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Rorb", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Tafa2", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Brinp3", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Pld5", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Kcnh5", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Sntg1", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Mlip", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Trhde", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Trpm3", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Kcnh1", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Dach1", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Pdzrn3", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Rasgrf2", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)
ImageFeaturePlot(obj.ctx.glutamatergic.COL, "Sgcd", size = 2.5, cols = c("lightgrey", "firebrick1"), max.cutoff = 1)

```


```{r}

LabelCellsByGeneFraction <- function(seurat_obj, genes, fraction, label, column_name) {
  # Ensure genes are present in the object
  genes_present <- intersect(genes, rownames(seurat_obj@assays$RNA@counts))
  if (length(genes_present) == 0) stop("No genes from the provided list are present in the Seurat object.")
  
  # Count how many of the specified genes are expressed in each cell
  gene_counts <- Matrix::colSums(seurat_obj@assays$RNA@counts[genes_present, ] > 0)
  
  # Determine the threshold based on the fraction
  threshold <- length(genes_present) * fraction
  
  # Label cells that express at least the threshold fraction of genes
  seurat_obj@meta.data[[column_name]] <- ifelse(gene_counts >= threshold, label, NA)
  
  return(seurat_obj)
}

L4.genes <- c("Rorb", "Thsd7a", "Sema3e", "Tafa2", "Brinp3", "Pld5", "Sntg1", "Mlip", "Trpm3", "Kcnh1", "Dach1", "Pdzrn3")

obj.ctx.glutamatergic.COL <- LabelCellsByGeneFraction(obj.ctx.glutamatergic.COL, L4.genes, 0.15, "L4", "L4-like")

```


```{r}

Idents(obj.ctx.glutamatergic.COL) <- "L4-like"
ImageDimPlot(obj.ctx.glutamatergic.COL, fov = "COL", size = 3)

```


```{r}

anchors <- FindTransferAnchors(reference = obj.glutamatergic, query = obj.ctx.glutamatergic.COL, normalization.method = "SCT")

predictions.assay <- TransferData(anchorset = anchors, refdata = obj.glutamatergic$subclass, prediction.assay = TRUE,
                                  weight.reduction = obj.ctx.glutamatergic.COL[["pca"]], dims = 1:30)

obj.ctx.glutamatergic.COL[["predictions"]] <- predictions.assay

```





























