---
title: "R Notebook"
output: html_notebook
---


```{r}

library(Seurat)
library(SeuratDisk)
library(reticulate)
library(scrubletR)
library(ggplot2)
library(patchwork)
library(dplyr)
library(data.table)
library(clustree)
library(reshape2)
library(tidyr)
library(gridExtra)
library(stringr)
library(plyr)
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/tools/seurat_functions.R")
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/tools/seurat_integration_functions.R")
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/xgboost/xgboost_train.R")
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/xgboost/plottingFxns.R")
colors_list <- GetColors()

data.path <- "E:/STOmics/seurat/Opossum/raw/"
obj.data <- Read10X(data.path, gene.column = 1)
obj.section <- CreateSeuratObject(counts = obj.data, project = "Opossum_V1_Stereo-seq")

mapping_path <- "E:/_genomes/Opossum_Mouse_Genes.txt"
obj.section <- MapGenes(obj, mapping_path, use_ids = TRUE)

obj.section$species <- "Opossum"
X <- read.csv(paste(data.path, "x_coords.csv", sep = ""), header = FALSE)
X <- (X - max(X)) * -1
obj.section <- AddMetaData(obj, X$V1, "X")
Y <- read.csv(paste(data.path, "y_coords.csv", sep = ""), header = FALSE)
Y <- (Y - max(Y)) * -1
obj.section <- AddMetaData(obj, Y$V1, "Y")

cents.df <- data.frame(X, Y)
colnames(cents.df) <- c("X", "Y")
rownames(cents.df) <- colnames(obj)
cents <- CreateCentroids(cents.df)
fov <- CreateFOV(
  cents, 
  type = "centroids",
  assay = "RNA",
  key = Key("FOV", quiet = TRUE)
)

section.cells <- read.csv("E:/STOmics/seurat/Opossum/selected_barcodes.csv", header = FALSE)

fov <- fov[as.character(section.cells$V1)]

obj.section[["FOV"]] <- fov

```


```{r, fig.width=10}

ImageDimPlot(obj.section, cols = "red")
ImageFeaturePlot(obj.section, "nCount_RNA", max.cutoff = 1000)
# ImageFeaturePlot(obj, "", max.cutoff = 10, axes = TRUE)

```


```{r}

obj.section <- ClusterSCT(obj.section, 0.1)

```


```{r, fig.width=10}

ImageDimPlot(obj.section,  group.by = "SCT_snn_res.0.1", size = 2)

```


```{r}

obj.glutamatergic <- readRDS("E:/Transcriptomics_V1/Opossum/seurat/opossum_v1_glutamatergic_processed.rds")
obj.glutamatergic$class <- "glutamatergic"
obj.gabaergic <- readRDS("E:/Transcriptomics_V1/Opossum/seurat/opossum_v1_gabaergic_processed.rds")
obj.gabaergic$class <- "gabaergic"
obj.nonneuronal <- readRDS("E:/Transcriptomics_V1/Opossum/seurat/opossum_v1_nonneuronal_processed.rds")

obj <- merge(obj.glutamatergic, y = c(obj.gabaergic, obj.nonneuronal))

```


```{r}

cell_mask <- Reduce(intersect, list(WhichCells(obj.section, expression = nFeature_RNA > 50),
                                    WhichCells(obj.section, expression = nCount_RNA > 100)))
  
gene_mask <- rownames(obj.section)[Matrix::rowSums(obj.section[["RNA"]]@counts > 0) > 8]

obj.section <- subset(obj.section, features = gene_mask, cells = cell_mask)

```


```{r}

obj$method <- "snRNA-seq"
obj.section$method <- "Stereo-seq"
obj.i <- IntegrateObjects(obj, obj.section, resolutions = c(0.5), subsample = TRUE)

```















