---
title: "Opossum V1 All"
output: 
---

```{r}

library(Seurat)
library(SeuratDisk)
library(reticulate)
library(scrubletR)
library(ggplot2)
library(patchwork)
library(dplyr)
library(data.table)
library(tidyverse)
source("C:/Ryan/GitHub/trachtenberg-lab/transcriptomics/tools/seurat_functions.R")

data_path <- "E:/Transcriptomics_V1/Opossum/output/Extended_lt10kb_ex5kb/"
sample_IDs <- c('OpossumV1-3A', 'OpossumV1-3B', 'OpossumV1-4A', 'OpossumV1-4B')
mapping_path <- "E:/_genomes/Opossum_Mouse_Genes.txt"

data <- PreprocessData(sample_IDs, data_path, "Opossum_V1", mapping_path)

obj.opossum <- data$obj
obj.opossum$species <- "Opossum"

```


```{r, fig.height=3, fig.width=4}

# Plot nFeature and nCount
PlotQC(data)

```


```{r}

# initial clustering
obj.opossum <- ClusterSCT(obj.opossum, c(1))
PlotClusters(obj.opossum, save.plots = TRUE)

```


```{r, fig.}

# remove obvious doublets, cluster again
obj.opossum <- subset(obj.opossum, idents = c(NA), invert = TRUE)
obj.opossum <- ClusterSCT(obj.opossum, c(1))
PlotClusters(obj.opossum)

```


```{r, fig.width=12, fig.height=6}

canon.markers <- list(
                    class = c("Snap25", "Sv2b", "Gad1", "Gad2"),
                    L23 = c("Cux2", "Ccbe1", "Mdga1", "Stard8"),
                    L4 = c("Whrn", "Rorb"),
                    L5IT = c("Bcl11b"),
                    L5NP = c("Nxph1", "Tshz2", "Trhr", "Slc17a8"),
                    L5PT = c("Bcl6", "Erg", "Reln"),
                    L6CT = c("Foxp2", "Syt6"),
                    L6IT = c("Zfp804b", "Cdh9"),
                    L6b = c("Ctgf", "Inpp4b", "Svil"),
                    Pvalb = c("Pvalb", "Myo5b"), 
                    Sst = c("Sst", "Chodl"),
                    Vip = c("Vip"),
                    Lamp5 = c("Lamp5", "Sv2c"),
                    Other = c("Frem1", "Stac", "Sncg", "Meis2"),
                    Astro = c("Aldh1l1", "Grin2c", "Phkg1"),
                    Endo = c("Pecam1", "Mecom", "Flt1"),
                    Micro = c("Cx3cr1", "Arhgap45", "Inpp5d"),
                    OD = c("Enpp6", "Mog", "St18"),
                    OPC = c("Pdgfra", "Cacng4"),
                    VLMC = c("Slc6a13", "Slc47a1", "Bnc2")
                      )

# look for canonical markers
DefaultAssay(obj.opossum) <- "RNA"
# PlotFeatures(obj.opossum, canon.markers)
DotPlot(obj.opossum, features = canon.markers, cols = c("lightgrey", "red"), scale = FALSE) + 
                         theme(axis.text.x = element_text(angle = 90, vjust = 0.25)) + 
                         theme(panel.background = element_rect(fill = "white", color = NA), plot.background = element_rect(fill = "white", color = NA)) + NoLegend()

```


```{r, fig.width=1.5, fig.height=6}

# look for canonical markers
DefaultAssay(obj.opossum) <- "RNA"
# PlotFeatures(obj.opossum, canon.markers)
p <- DotPlot(obj.opossum, features = c("Snap25", "Sv2b", "Gad1", "Gad2"), cols = c("lightgrey", "red"), scale = FALSE) + 
                         theme(axis.text.x = element_text(angle = 90, vjust = 0.25)) + 
                         theme(panel.background = element_rect(fill = "white", color = NA), plot.background = element_rect(fill = "white", color = NA)) + NoLegend()

ggsave("E:/Opossum_Paper/Figure S1/Class_DotPlot.svg", plot = p)

```


```{r}

DefaultAssay(obj.opossum) <- "SCT"
markers.22 <- FindMarkers(obj.opossum, ident.1 = 22, ident.2 = c(6, 16, 18, 21, 23, 25))

```


```{r}

DefaultAssay(obj.opossum) <- "SCT"
ident = "SCT_snn_res.1"
gene <- "Gad1"
DimPlot(obj.opossum, reduction = "umap", group.by = ident, label = TRUE, raster = FALSE) + NoLegend() + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
FeaturePlot(obj.opossum, gene, raster = FALSE) + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
VlnPlot(obj.opossum, gene, group.by = ident)

```


```{r}

cells.glutamatergic <- WhichCells(obj.opossum, ident = c(1, 4, 5, 8, 9, 10, 11, 13, 14, 15, 17, 20, 24))
obj.opossum$class <- ifelse(colnames(obj.opossum) %in% cells.glutamatergic, "glutamatergic", "ambiguous")

cells.gabaergic <- WhichCells(obj.opossum, ident = c(6, 16, 18, 21, 22, 23, 25))
obj.opossum$class[cells.gabaergic] <- "gabaergic"

cells.nonneuronal <- WhichCells(obj.opossum, ident = c(2, 3, 7, 12, 19, 26, 27))
obj.opossum$class[cells.nonneuronal] <- "nonneuronal"

```


```{r}

DimPlot(obj.opossum, reduction = "umap", group.by = "class", label = FALSE, raster = FALSE, shuffle = TRUE) + xlim(-18, 18) + ylim(-18, 18) + coord_equal()

```


```{r}

p <- DimPlot(obj.opossum, reduction = "umap", group.by = "class", label = FALSE, raster = FALSE, shuffle = TRUE) + xlim(-18, 18) + ylim(-18, 18) + coord_equal()

ggsave("E:/Opossum_Paper/Figure S1/Class_DimPlot.svg", plot = p)

```


```{r}

obj.opossum$all_SCT_snn_res.1 <- obj.opossum$SCT_snn_res.1
Idents(obj.opossum) <- "class"

obj.opossum.glutamatergic <- subset(obj.opossum, idents = "glutamatergic")
saveRDS(obj.opossum.glutamatergic, "E:/Transcriptomics_V1/Opossum/seurat/opossum_v1_glutamatergic.rds")

obj.opossum.gabaergic <- subset(obj.opossum, idents = "gabaergic")
saveRDS(obj.opossum.gabaergic, "E:/Transcriptomics_V1/Opossum/seurat/opossum_v1_gabaergic.rds")

obj.opossum.nonneuronal <- subset(obj.opossum, idents = "nonneuronal")
saveRDS(obj.opossum.nonneuronal, "E:/Transcriptomics_V1/Opossum/seurat/opossum_v1_nonneuronal.rds")

```

