---
title: "R Notebook"
output: html_notebook
---


```{r, fig.width = 6, fig.height = 5}

library(ggplot2)
library(dplyr)

# Load the files and add a 'sample' column
files <- list(
  # "E:/bams/opossum/dists_3A.bed" = "3A",
  # "E:/bams/opossum/dists_3B.bed" = "3B",
  "E:/bams/opossum/dists_4A.bed" = "4A"
  # "E:/bams/opossum/dists_4B.bed" = "4B"
)

# Read and combine all files
closest.bed <- bind_rows(lapply(names(files), function(path) {
  df <- read.table(path, sep = "\t", header = FALSE)
  df$sample <- files[[path]]
  return(df)
}))

# Plot
p <- ggplot(closest.bed, aes(x = V13, fill = sample, color = sample)) +
  geom_histogram(
    binwidth = 1000, 
    boundary = 0,
    aes(y = ..count.. / tapply(..count.., ..PANEL.., sum)[..PANEL..] * 100),
    alpha = 0.4,
    position = "identity"
  ) +
  scale_x_continuous(
    limits = c(-50000, 50000),
    oob = scales::squish
  ) +
  scale_fill_brewer(palette = "Set1") +
  scale_color_brewer(palette = "Set1") +
  labs(
    x = "Closest Gene (bp)",
    y = "Percentage",
    fill = "Sample",
    color = "Sample"
  ) +
  theme_minimal()

```


```{r}

ggsave("E:/Opossum_Paper/Figure S1/Intergenic_Reads_Dist_Opossum.svg", plot = p, width = 6, height = 5)

```


```{r}

p <- ggplot(closest.bed, aes(x = V13, fill = sample, color = sample)) +
  geom_histogram(
    binwidth = 1000, 
    boundary = 0,
    aes(y = ..count.. / sum(..count..) * 100),  # per-facet percentage
    fill = "lightblue", 
    color = "black"
  ) +
  scale_x_continuous(
    limits = c(-50000, 50000),
    oob = scales::squish
  ) +
  labs(
    x = "Closest Gene (bp)",
    y = "Percentage"
  ) +
  facet_wrap(~ sample, ncol = 2) +  # 2x2 layout
  theme_minimal()

```


```{r}

ggsave("E:/Opossum_Paper/Figure S1/Intergenic_Reads_Dist_Opossum_Grid.svg", plot = p, width = 6, height = 5)

```


```{r, fig.width=6, fig.height=5}

ggplot(closest.bed, aes(x = V13, fill = sample, color = sample)) +
  geom_histogram(
    binwidth = 1000, 
    boundary = 0,
    aes(y = ..count.. / sum(..count..) * 100),  # per-facet percentage
    fill = "lightblue", 
    color = "black"
  ) +
  scale_x_continuous(
    limits = c(-10000, 10000),
    oob = scales::squish
  ) +
  scale_y_continuous(
    limits = c(0, 6)
  ) +
  labs(
    x = "Closest Gene (bp)",
    y = "Percentage"
  ) +
  facet_wrap(~ sample, ncol = 2) +  # 2x2 layout
  theme_minimal()

```


```{r}

ggsave("E:/Opossum_Paper/Figure S1/Intergenic_Reads_Dist_4A_Zoom.svg", plot = p, width = 6, height = 5)

```


```{r}

# filter by gene biotype
get.var <- function(var.col, item) {
  
  var.col.spl <- strsplit(variable.col, "; ")
  var.col.idx <- lapply(var.col.spl, function(x) grepl(item, x, fixed = TRUE))
  var.col.item <- mapply(function(x, y) x[y], var.col.spl, var.col.idx)
  missing <- lapply(var.col.item, function(x) identical(x, character(0)))
  var.col.item[unlist(missing)] <- NA
  var.col.nona <- lapply(var.col.item[is.na(var.col.item) == F], function(x) strsplit(x, " ")[[1]][2])
  var.col.nona <- lapply(var.col.nona, function(x) strsplit(x, ";")[[1]])
  var.col.item[is.na(var.col.item) == F] <- var.col.nona
  return(unlist(var.col.item))
  
}

gtf = read.table("E:/_genomes/Opossum/Monodelphis_domestica.ASM229v1.112.gtf", skip = 5, sep = "\t", header = F)

colnames(gtf) <- c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute")
variable.col <- gtf$attribute
gtf <- gtf[, 1:8]

gtf$gene.id <- get.var(variable.col, "gene_id")
gtf$gene.name <- get.var(variable.col, "gene_name")
gtf$gene.biotype <- get.var(variable.col, "gene_biotype")

```


```{r}

library(ggplot2)

pc.ids <- gtf$gene.id[gtf$feature == "gene" & gtf$gene.biotype == "protein_coding"]
closest.bed.pc <- closest.bed[closest.bed$V10 %in% pc.ids,]

# Plot histogram using ggplot2 with specified settings
ggplot(closest.bed.pc, aes(x = V13)) +
  geom_histogram(
    binwidth = 1000, 
    fill = "lightblue", 
    color = "black",
    aes(y = ..count.. / sum(..count..) * 100)
  ) +
  scale_x_continuous(
    limits = c(-50000, 50000),
    oob = scales::squish
  ) +
  labs(
    title = "Intergenic Reads",
    x = "Closest protein_coding Gene (bp)",
    y = "Percentage"
  ) +
  theme_minimal()

```


```{r}

library(ggplot2)

lnc.ids <- gtf$gene.id[gtf$feature == "gene" & gtf$gene.biotype == "lncRNA"]
closest.bed.lnc <- closest.bed[closest.bed$V10 %in% lnc.ids,]

# Plot histogram using ggplot2 with specified settings
ggplot(closest.bed.lnc, aes(x = V13)) +
  geom_histogram(
    binwidth = 1000, 
    fill = "lightblue", 
    color = "black",
    aes(y = ..count.. / sum(..count..) * 100)
  ) +
  scale_x_continuous(
    limits = c(-50000, 50000),
    oob = scales::squish
  ) +
  labs(
    title = "Intergenic Reads",
    x = "Closest lncRNA Gene (bp)",
    y = "Percentage"
  ) +
  theme_minimal()

```

